checker/src/main/java/org/checkerframework/checker/nullness/NullnessTransfer.java

```diff
@@ -357,21 +358,35 @@ public TransferResult<NullnessValue, NullnessStore> visitMethodInvocation(
             MethodInvocationNode n, TransferInput<NullnessValue, NullnessStore> in) {
         TransferResult<NullnessValue, NullnessStore> result = super.visitMethodInvocation(n, in);
 
-        // Make receiver non-null.
+        MethodInvocationTree tree = n.getTree();
+        ExecutableElement method = TreeUtils.elementFromUse(tree);
+        boolean nonNullAssumption =
+                !analysis.getTypeFactory()
+                        .getChecker()
+                        .hasOption("receiverAndParamsNullableAfterMethodCall");
+
+        boolean isMethodSideEffectFree = PurityUtils.isSideEffectFree(atypeFactory, method);
         Node receiver = n.getTarget().getReceiver();
-        makeNonNull(result, receiver);
+        if (nonNullAssumption
+                || JavaExpression.fromNode(receiver).isUnassignableByOtherCode()
+                || isMethodSideEffectFree) {
+            // Make receiver non-null.
+            makeNonNull(result, receiver);
+        }
 
         // For all formal parameters with a non-null annotation, make the actual argument non-null.
         // The point of this is to prevent cascaded errors -- the Nullness Checker will issue a
         // warning for the method invocation, but not for subsequent uses of the argument.  See test
         // case FlowNullness.java.
-        MethodInvocationTree tree = n.getTree();
-        ExecutableElement method = TreeUtils.elementFromUse(tree);
         AnnotatedExecutableType methodType = nullnessTypeFactory.getAnnotatedType(method);
         List<AnnotatedTypeMirror> methodParams = methodType.getParameterTypes();
         List<? extends ExpressionTree> methodArgs = tree.getArguments();
         for (int i = 0; i < methodParams.size() && i < methodArgs.size(); ++i) {
-            if (methodParams.get(i).hasAnnotation(NONNULL)) {
+            if (methodParams.get(i).hasAnnotation(NONNULL)
+                    && (nonNullAssumption
+                            || isMethodSideEffectFree
+                            || JavaExpression.fromTree(methodArgs.get(i))
+                                    .isUnassignableByOtherCode())) {
```

Comment:
Should `visitMethodAccess` https://github.com/eisop/checker-framework/blob/c62ed408a5a957b86457c776f4aeaf588f789f00/checker/src/main/java/org/checkerframework/checker/nullness/NullnessTransfer.java#L328
also be changed?
Can you add all the test cases we wrote to the PR and make sure `visitMethodAccess` is also covered?

Reply:
It should be changed, but I noticed that a `MethodAccessNode`, will always be followed by a `MethodInvocationNode`. Thus, even though here it is non-null, later `visitMethodInvocation` will make it nullable if the method is impure. That maybe the reason why our tests still pass

Reply:
This line of code should not be deleted. This is an unconditional makeNonNull, as dereferencing requires the receiver to be non-null.